File: algorithms/code/leetcode/lc172_factorial_trailing_zeroes/lc172_factorial_trailing_zeroes.py
# https://leetcode.com/problems/factorial-trailing-zeroes/

import unittest


class Solution:
    def trailingZeroes(self, n: int) -> int:
        step = 5
        count = 0
        while step <= n:
            count += n // step
            step = step * 5
        return count


class TestCode(unittest.TestCase):
    def test_5(self) -> None:
        assert Solution().trailingZeroes(5) == 1

    def test_1808548329(self) -> None:
        assert Solution().trailingZeroes(1808548329) == 452137076

    def test_2147483647(self) -> None:
        assert Solution().trailingZeroes(2147483647) == 536870902

File: algorithms/code/leetcode/lc172_factorial_trailing_zeroes/lc172_factorial_trailing_zeroes.rs
// https://leetcode.com/problems/factorial-trailing-zeroes/
// #medium

pub struct Solution;

impl Solution {
    pub fn trailing_zeroes(n: i32) -> i32 {
        let mut step: i64 = 5;
        let mut count: i64 = 0;
        while step <= n as i64 {
            count += n as i64 / step;
            step *= 5;
        }
        count as i32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_5() {
        assert_eq!(Solution::trailing_zeroes(5), 1);
    }

    #[test]
    fn test_1808548329() {
        assert_eq!(Solution::trailing_zeroes(1808548329), 452137076);
    }

    #[test]
    fn test_2147483647() {
        assert_eq!(Solution::trailing_zeroes(2147483647), 536870902);
    }
}
